scylla/src/transport/load_balancing/default.rs

```diff
@@ -411,6 +432,80 @@ impl DefaultPolicy {
+    // caused by Paxos conflicts, we always try to query replicas in the same, ring order.
+    //
+    // If preferred rack and DC are set, then the first (encountered on the ring) replica
+    // that resides in that rack in that DC **and** satisfies the `predicate`  is returned.
+    //
+    // If preferred DC is set, then the first (encountered on the ring) replica
+    // that resides in that DC **and** satisfies the `predicate` is returned.
+    //
+    // If no DC/rack preferences are set, then the only possible replica to be returned
+    // (due to expensive computation of the others, and we avoid expensive computation in `pick()`)
+    // is the primary replica. It is returned **iff** it satisfies the predicate, else None.
+    fn pick_first_replica<'a>(
+        &'a self,
+        ts: &TokenWithStrategy<'a>,
+        replica_location: ReplicaLocationCriteria<'a>,
+        predicate: &'a impl Fn(&NodeRef<'a>) -> bool,
+        cluster: &'a ClusterData,
+    ) -> Option<NodeRef<'a>> {
+        match replica_location {
+            ReplicaLocationCriteria::Any => {
+                // ReplicaSet returned by ReplicaLocator for this case:
+                // 1) can be precomputed and lated used cheaply,
+                // 2) returns replicas in the **non-ring order** (this because ReplicaSet chains
+                //    ring-ordered replicas sequences from different DCs, thus not preserving
+                //    the global ring order).
+                // Because of 2), we can't use a precomputed ReplicaSet, but instead we need ReplicasOrdered.
+                // As ReplicasOrdered can compute cheaply only the primary global replica
+                // (computation of the remaining ones is expensive), in case that the primary replica
+                // does not satisfy the `predicate`, None is returned. All expensive computation
+                // is to be done only when `fallback()` is called.
+                self.nonfiltered_replica_set(ts, replica_location, cluster)
+                    .into_replicas_ordered()
+                    .into_iter()
+                    .next()
+                    .and_then(|primary_replica| {
+                        predicate(&primary_replica).then_some(primary_replica)
+                    })
+            }
+            ReplicaLocationCriteria::Datacenter(_)
+            | ReplicaLocationCriteria::DatacenterAndRack(_, _) => {
+                // ReplicaSet returned by ReplicaLocator for this case:
+                // 1) can be precomputed and lated used cheaply,
+                // 2) returns replicas in the ring order (this is not true for the case
+                //    when multiple DCs are allowed, because ReplicaSet chains replicas sequences
+                //    from different DCs, thus not preserving the global ring order)
+                self.replicas(
+                    ts,
+                    replica_location,
+                    move |node| predicate(node),
```

Comment:
What about rack predicate?

Reply:
`replicas()` is responsible for calling `make_rack_predicate()`.